### PR TITLE
Update to latest Mutter API

### PR DESCRIPTION
**Pull request by danrabbit**

### CODE REVIEW

File: src/Objects/Monitor.vala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                if (mode.is_current) {
+                    return mode;
+                }
+            }
+
+            return null;
+        }
+    }
+
+    public Display.MonitorMode preferred_mode {
```
Comment (donadigo) [inline src/Objects/Monitor.vala]: Should be probably nullable `MonitorMode?`

File: src/Objects/MonitorManager.vala [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        clone_virtual_monitor.monitors.add_all (monitors);
+        var modes = clone_virtual_monitor.get_available_modes ();
+        /*
+         * Two choices here:
+         *  - Use the largest resolution already in use.
+         *  - Fallback to the largest resultion available.
+         */
+
+        Display.MonitorMode? largest_mode_in_use = null;
+        Display.MonitorMode largest_mode = modes.get (0);
```
Comment (donadigo) [inline src/Objects/MonitorManager.vala]: Check if `modes` is not empty (?)

File: src/Objects/MonitorManager.vala [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+                current_mode.is_current = false;
+                preferred_mode.is_current = true;
+                single_virtual_monitor.scale = preferred_mode.preferred_scale;
+            }
+
+            single_virtual_monitor.monitors.add (monitor);
+            new_virtual_monitors.add (single_virtual_monitor);
+        }
+
+        new_virtual_monitors.get(0).primary = true;
```
Comment (donadigo) [inline src/Objects/MonitorManager.vala]: Check if `new_virtual_monitors` is not empty.
Also you can use the `[0]` syntax with Gee.

File: src/Objects/MonitorMode.vala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    public string id { get; set; }
+    public int width { get; set; }
+    public int height { get; set; }
+    public double frequency { get; set; }
+    public double preferred_scale { get; set; }
+    public double[] supported_scales { get; set; }
+    public bool is_preferred { get; set; }
+    public bool is_current { get; set; }
+
+    private string resolution_cache = null;
```
Comment (donadigo) [inline src/Objects/MonitorMode.vala]: Make this nullable.

File: src/Utils.vala [hunk 5]
```diff
@@ -34,4 +34,49 @@ namespace Display.Utils {
...
+
+                first_monitor = false;
+            } else {
+                var to_remove = new Gee.LinkedList<Display.MonitorMode> ();
+                foreach (var mode_to_check in common_modes) {
+                    bool mode_found = false;
+                    foreach (var monitor_mode in monitor.modes) {
+                        if (mode_to_check.width == monitor_mode.width &&
+                            mode_to_check.height == monitor_mode.height) {
+                            mode_found = true;
```
Comment (donadigo) [inline src/Utils.vala]: `break` here. There's no need to search for more modes if there's already one matching.

File: src/Objects/Monitor.vala (hunk 1 above)
Comment (tintou) [reply src/Objects/Monitor.vala]: I would prefer to probably return `modes.get (0)` instead

File: src/Objects/MonitorManager.vala (hunk 2 above)
Comment (tintou) [reply src/Objects/MonitorManager.vala]: Let's assume that modes is never empty

File: src/Objects/MonitorManager.vala (hunk 3 above)
Comment (tintou) [reply src/Objects/MonitorManager.vala]: new_virtual_monitors can't be empty

File: src/Objects/MonitorMode.vala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            case 17:
+                aspect = "16∶9";
+                break;
+            case 23:
+                aspect = "21∶9";
+                break;
+            case 12:
+                aspect = "5∶4";
+                break;
+                /* This catches 1.5625 as well (1600x1024) when maybe it shouldn't. */
```
[the remark this answers is not in the document]
Comment (tintou) [reply src/Objects/MonitorMode.vala]: This is from the old plug, I don't want to modify it in this merge request

File: src/Utils.vala (hunk 5 above)
Comment (tintou) [reply src/Utils.vala]: thanks, indeed. I'm always afraid of using `break` inside nested loops, I miss the `break n` syntax

**Pull request merged**